pallets/xyk-liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,927 @@
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::genesis_config]
+	#[cfg_attr(feature = "std", derive(Default))]
+	pub struct GenesisConfig {}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			let pallet_account = <Pallet<T>>::account_id();
+
+			T::NonDustableWhitelistHandler::add_account(&pallet_account).unwrap();
+
+			T::NFTHandler::create_typed_collection(
+				pallet_account,
+				T::NftCollectionId::get(),
+				CollectionType::LiquidityMining,
+			)
+			.unwrap();
+		}
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + TypeInfo {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Currency for transfers.
+		type MultiCurrency: MultiCurrency<Self::AccountId, CurrencyId = AssetId, Balance = Balance>;
+
+		/// AMM helper functions.
+		type AMM: AMM<Self::AccountId, AssetId, AssetPair, Balance>;
+
+		/// The origin account that can create new liquidity mining program.
+		type CreateOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Pallet id.
+		type PalletId: Get<PalletId>;
+
+		/// The block number provider
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+
+		/// NFT collection id for liq. mining deposit nfts. Has to be within the range of reserved NFT class IDs.
+		#[pallet::constant]
+		type NftCollectionId: Get<primitives::CollectionId>;
+
+		/// Non fungible handling
+		type NFTHandler: Mutate<Self::AccountId>
+			+ Create<Self::AccountId>
+			+ Inspect<Self::AccountId, CollectionId = primitives::CollectionId, ItemId = DepositId>
+			+ CreateTypedCollection<Self::AccountId, primitives::CollectionId, CollectionType>;
+
+		/// Liquidity mining handler for managing liquidity mining functionalities
+		type LiquidityMiningHandler: LiquidityMiningMutate<
+			Self::AccountId,
+			AssetId,
+			BlockNumberFor<Self>,
+			Error = DispatchError,
+			AmmPoolId = Self::AccountId,
+			Balance = Balance,
+			LoyaltyCurve = LoyaltyCurve,
+			Period = PeriodOf<Self>,
+		>;
+
+		/// Account whitelist manager to exclude pool accounts from dusting mechanism.
+		type NonDustableWhitelistHandler: DustRemovalAccountWhitelist<Self::AccountId, Error = DispatchError>;
+
+		/// Weight information for extrinsic in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Nft pallet didn't return an owner.
+		CantFindDepositOwner,
+
+		/// Account balance of amm pool shares is not sufficient.
+		InsufficientAmmSharesBalance,
+
+		/// AMM pool does not exist
+		AmmPoolDoesNotExist,
+
+		/// Account is not deposit owner.
+		NotDepositOwner,
+
+		/// AMM did not return assets for given `amm_pool_id`
+		CantGetAmmAssets,
+
+		/// Yield farm can not be found
+		YieldFarmNotFound,
+
+		///Deposit data not found
+		DepositDataNotFound,
+
+		/// Calculated reward to claim is 0.
+		ZeroClaimedRewards,
+	}
```

Comment:
Missing tests for `CantFindDepositOwner`, `CantGetAmmAssets`, `YieldFarmNotFound`